[unused_declaration] Unused class/struct is not detected if it conforms to a protocol in extension

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When a `class`/`struct` is not used anywhere but has conformance to a protocol in extension, it's not detected as unused by `unused_declaration` analyzer rule, while when conformance to a protocol is in declaration, it is detected as unused.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
# // Foo.swift
#
# protocol Foo {}

# // FooImpl.swift
# 
# final class FooImpl {}
#
# extension FooImpl: Foo {}

$ mint run realm/SwiftLint@0.54.0 analyze --compiler-log-path outputs/xcodebuild.log

Analyzing Swift files in current working directory
Collecting 'FooImpl.swift' (1/2)
Collecting 'Foo.swift' (2/2)
Analyzing 'FooImpl.swift' (1/2)
Analyzing 'Foo.swift' (2/2)
Done analyzing! Found 0 violations, 0 serious in 2 files.

# // Goo.swift
#
# protocol Goo {}

# // GooImpl.swift
# 
# final class GooImpl: Goo {}

$ mint run realm/SwiftLint@0.54.0 analyze --compiler-log-path outputs/xcodebuild.log

Analyzing Swift files in current working directory
Collecting 'Goo.swift' (1/2)
Collecting 'GooImpl.swift' (2/2)
Analyzing 'Goo.swift' (1/2)
Analyzing 'GooImpl.swift' (2/2)
/Users/vakhid.betrakhmadov/Developer/swift-compilation-database/Sources/ModuleA/GooImpl.swift:1:13: error: Unused Declaration Violation: Declarations should be referenced at least once within all files linted (unused_declaration)
Done analyzing! Found 1 violation, 1 serious in 2 files.
make: *** [swiftlint] Error 2
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?

0.54.0

* Installation method used (Homebrew, CocoaPods, building from source, etc)?

Homebrew, mint

* Paste your configuration file:

```yml
analyzer_rules:
  - unused_declaration

unused_declaration:
  include_public_and_open: false
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents.

No

* Which Xcode version are you using (check `xcodebuild -version`)?

Xcode 15.2
Build version 15C500b

* Do you have a sample that shows the issue?

https://github.com/vakhidbetrakhmadov/swift-compilation-database

```bash
$ make swiftlint

rm -rf Example.xcodeproj
rm -rf outputs
mkdir outputs
mint run yonaskolb/XcodeGen@2.40.1
⚙️  Generating plists...
⚙️  Generating project...
⚙️  Writing project...
Created project at /Users/vakhid.betrakhmadov/Developer/swift-compilation-database/Example.xcodeproj
xcodebuild \
                -project Example.xcodeproj \
                -scheme "ModuleA" \
                -destination generic/platform=MacOS \
                clean build \
                > outputs/xcodebuild.log
mint run realm/SwiftLint@0.54.0 analyze --compiler-log-path outputs/xcodebuild.log
Analyzing Swift files in current working directory
Collecting 'FooImpl.swift' (1/4)
Collecting 'Goo.swift' (2/4)
Collecting 'Foo.swift' (3/4)
Collecting 'GooImpl.swift' (4/4)
Analyzing 'FooImpl.swift' (1/4)
Analyzing 'Goo.swift' (2/4)
Analyzing 'Foo.swift' (3/4)
Analyzing 'GooImpl.swift' (4/4)
/Users/vakhid.betrakhmadov/Developer/swift-compilation-database/Sources/ModuleA/GooImpl.swift:1:13: error: Unused Declaration Violation: Declarations should be referenced at least once within all files linted (unused_declaration)
Done analyzing! Found 1 violation, 1 serious in 4 files.
make: *** [swiftlint] Error 2
```

This sample project demonstrates how a `class`/`struct` that is not used anywhere but has conformance to a protocol in extension is not detected as unused by `unused_declaration` analyzer rule (`FooImpl`), while a `class`/`struct` that is not used anywhere but conforms to a protocol in its declaration is (`GooImpl`).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -223,6 +223,11 @@
   [SimplyDanny](https://github.com/SimplyDanny)
   [#5489](https://github.com/realm/SwiftLint/pull/5489)
 
+* Ensure that declarations referenced only as extended types do not count as
+  used by means of the `unused_declaration` rule.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#5550](https://github.com/realm/SwiftLint/issues/5550)
+
 * Fix some false positives in `multiline_literal_brackets` rule that would
   happen when comments are present.  
   [Marcelo Fabri](https://github.com/marcelofabri)

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/CaptureVariableRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/CaptureVariableRule.swift
@@ -206,13 +206,13 @@ private extension SwiftLintFile {
         let offsets = self.captureListVariableOffsets()
         guard !offsets.isEmpty, let indexEntities = index(compilerArguments: compilerArguments) else { return Set() }
 
-        return Set(indexEntities.traverseEntitiesDepthFirst {
+        return Set(indexEntities.traverseEntitiesDepthFirst { _, entity in
             guard
-                let kind = $0.kind,
+                let kind = entity.kind,
                 kind.hasPrefix("source.lang.swift.ref.var."),
-                let usr = $0.usr,
-                let line = $0.line,
-                let column = $0.column,
+                let usr = entity.usr,
+                let line = entity.line,
+                let column = entity.column,
                 let offset = stringView.byteOffset(forLine: line, bytePosition: column)
             else { return nil }
             return offsets.contains(offset) ? CaptureVariableRule.Variable(usr: usr, offset: offset) : nil
@@ -245,16 +245,16 @@ private extension SwiftLintFile {
         let offsets = self.declaredVariableOffsets()
         guard !offsets.isEmpty, let indexEntities = index(compilerArguments: compilerArguments) else { return Set() }
 
-        return Set(indexEntities.traverseEntitiesDepthFirst {
+        return Set(indexEntities.traverseEntitiesDepthFirst { _, entity in
             guard
-                let declarationKind = $0.declarationKind,
+                let declarationKind = entity.declarationKind,
                 Self.checkedDeclarationKinds.contains(declarationKind),
-                let line = $0.line,
-                let column = $0.column,
+                let line = entity.line,
+                let column = entity.column,
                 let offset = stringView.byteOffset(forLine: line, bytePosition: column),
                 offsets.contains(offset)
             else { return nil }
-            return $0.usr
+            return entity.usr
         })
     }
 

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnusedDeclarationRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnusedDeclarationRule.swift
@@ -46,7 +46,7 @@ struct UnusedDeclarationRule: AnalyzerRule, CollectingRule {
         }
 
         return FileUSRs(
-            referenced: file.referencedUSRs(index: index),
+            referenced: file.referencedUSRs(index: index, editorOpen: editorOpen),
             declared: file.declaredUSRs(index: index,
                                         editorOpen: editorOpen,
                                         compilerArguments: compilerArguments,
@@ -89,11 +89,16 @@ private extension SwiftLintFile {
             .map(SourceKittenDictionary.init)
     }
 
-    func referencedUSRs(index: SourceKittenDictionary) -> Set<String> {
-        return Set(index.traverseEntitiesDepthFirst { _, entity -> String? in
+    func referencedUSRs(index: SourceKittenDictionary, editorOpen: SourceKittenDictionary) -> Set<String> {
+        return Set(index.traverseEntitiesDepthFirst { parent, entity -> String? in
             if let usr = entity.usr,
-                let kind = entity.kind,
-                kind.starts(with: "source.lang.swift.ref") {
+               let kind = entity.kind,
+               kind.starts(with: "source.lang.swift.ref"),
+               !parent.extends(reference: entity),
+               let line = entity.line,
+               let column = entity.column,
+               let nameOffset = stringView.byteOffset(forLine: line, bytePosition: column),
+               editorOpen.propertyAtOffset(nameOffset, property: \.kind) != "source.lang.swift.decl.extension" {
                 return usr
             }
 
@@ -133,7 +138,8 @@ private extension SwiftLintFile {
             return nil
         }
 
-        if !configuration.includePublicAndOpen, [.public, .open].contains(editorOpen.aclAtOffset(nameOffset)) {
+        if !configuration.includePublicAndOpen,
+           [.public, .open].contains(editorOpen.propertyAtOffset(nameOffset, property: \.accessibility)) {
             return nil
         }
 
@@ -230,14 +236,14 @@ private extension SourceKittenDictionary {
         return value["key.is_implicit"] as? Bool == true
     }
 
-    func aclAtOffset(_ offset: ByteCount) -> AccessControlLevel? {
+    func propertyAtOffset<T>(_ offset: ByteCount, property: KeyPath<Self, T?>) -> T? {
         if let nameOffset,
             nameOffset == offset,
-            let acl = accessibility {
-            return acl
+            let field = self[keyPath: property] {
+            return field
         }
         for child in substructure {
-            if let acl = child.aclAtOffset(offset) {
+            if let acl = child.propertyAtOffset(offset, property: property) {
                 return acl
             }
         }
@@ -295,6 +301,21 @@ private extension SourceKittenDictionary {
         ]
 
         return resultBuilderStaticMethods.contains(name)
+    }
+
+    func extends(reference other: Self) -> Bool {
+        if let kind, kind.starts(with: "source.lang.swift.decl.extension") {
+            let extendedKind = kind.components(separatedBy: ".").last
+            return extendedKind != nil && extendedKind == other.referencedKind
+        }
+        return false
+    }
+
+    private var referencedKind: String? {
+        if let kind, kind.starts(with: "source.lang.swift.ref") {
+            return kind.components(separatedBy: ".").last
+        }
+        return nil
     }
 }
 

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnusedDeclarationRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnusedDeclarationRule.swift
@@ -90,7 +90,7 @@ private extension SwiftLintFile {
     }
 
     func referencedUSRs(index: SourceKittenDictionary) -> Set<String> {
-        return Set(index.traverseEntitiesDepthFirst { entity -> String? in
+        return Set(index.traverseEntitiesDepthFirst { _, entity -> String? in
             if let usr = entity.usr,
                 let kind = entity.kind,
                 kind.starts(with: "source.lang.swift.ref") {
@@ -104,7 +104,7 @@ private extension SwiftLintFile {
     func declaredUSRs(index: SourceKittenDictionary, editorOpen: SourceKittenDictionary,
                       compilerArguments: [String], configuration: UnusedDeclarationConfiguration)
     -> Set<UnusedDeclarationRule.DeclaredUSR> {
-        return Set(index.traverseEntitiesDepthFirst { indexEntity in
+        return Set(index.traverseEntitiesDepthFirst { _, indexEntity in
             self.declaredUSR(indexEntity: indexEntity, editorOpen: editorOpen, compilerArguments: compilerArguments,
                              configuration: configuration)
         })
@@ -140,14 +140,14 @@ private extension SwiftLintFile {
         // Skip CodingKeys as they are used for Codable generation
         if kind == .enum,
             indexEntity.name == "CodingKeys",
-            case let allRelatedUSRs = indexEntity.traverseEntitiesDepthFirst(traverseBlock: { $0.usr }),
+            case let allRelatedUSRs = indexEntity.traverseEntitiesDepthFirst(traverseBlock: { $1.usr }),
             allRelatedUSRs.contains("s:s9CodingKeyP") {
             return nil
         }
 
         // Skip `static var allTests` members since those are used for Linux test discovery.
         if kind == .varStatic, indexEntity.name == "allTests" {
-            let allTestCandidates = indexEntity.traverseEntitiesDepthFirst { subEntity -> Bool in
+            let allTestCandidates = indexEntity.traverseEntitiesDepthFirst { _, subEntity -> Bool in
                 subEntity.value["key.is_test_candidate"] as? Bool == true
             }
 

--- a/Source/SwiftLintBuiltInRules/Rules/Lint/UnusedDeclarationRuleExamples.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Lint/UnusedDeclarationRuleExamples.swift
@@ -207,8 +207,23 @@ struct UnusedDeclarationRuleExamples {
         }
 
         _ = ComponentBuilder()
+        """),
+        Example("""
+        protocol ↓Foo {}
+        extension Foo {}
+        """),
+        Example("""
+        class ↓C<T> {}
+        extension C<Int> {}
         """)
-    ] + platformSpecificTriggeringExamples
+    ] + ["actor", "enum", "class", "struct"].map {
+        Example("""
+        protocol Foo {}
+        \($0) ↓FooImpl {}
+        extension FooImpl {}
+        extension FooImpl: Foo {}
+        """, excludeFromDocumentation: true)
+    } + platformSpecificTriggeringExamples
 
 #if os(macOS)
     private static let platformSpecificNonTriggeringExamples = [

--- a/Source/SwiftLintCore/Extensions/Dictionary+SwiftLint.swift
+++ b/Source/SwiftLintCore/Extensions/Dictionary+SwiftLint.swift
@@ -190,6 +190,9 @@ public struct SourceKittenDictionary {
 }
 
 extension SourceKittenDictionary {
+    /// Block executed for every encountered entity during traversal of a dictionary.
+    public typealias TraverseBlock<T> = (_ parent: SourceKittenDictionary, _ entity: SourceKittenDictionary) -> T?
+
     /// Traversing all substuctures of the dictionary hierarchically, calling `traverseBlock` on each node.
     /// Traversing using depth first strategy, so deepest substructures will be passed to `traverseBlock` first.
     ///
@@ -216,21 +219,20 @@ extension SourceKittenDictionary {
     /// Traversing all entities of the dictionary hierarchically, calling `traverseBlock` on each node.
     /// Traversing using depth first strategy, so deepest substructures will be passed to `traverseBlock` first.
     ///
-    /// - parameter traverseBlock: block that will be called for each entity in the dictionary.
+    /// - parameter traverseBlock: Block that will be called for each entity and its parent in the dictionary.
     ///
     /// - returns: The list of entity dictionaries with updated values from the traverse block.
-    public func traverseEntitiesDepthFirst<T>(traverseBlock: (SourceKittenDictionary) -> T?) -> [T] {
+    public func traverseEntitiesDepthFirst<T>(traverseBlock: TraverseBlock<T>) -> [T] {
         var result: [T] = []
         traverseEntitiesDepthFirst(collectingValuesInto: &result, traverseBlock: traverseBlock)
         return result
     }
 
-    private func traverseEntitiesDepthFirst<T>(collectingValuesInto array: inout [T],
-                                               traverseBlock: (SourceKittenDictionary) -> T?) {
+    private func traverseEntitiesDepthFirst<T>(collectingValuesInto array: inout [T], traverseBlock: TraverseBlock<T>) {
         entities.forEach { subDict in
             subDict.traverseEntitiesDepthFirst(collectingValuesInto: &array, traverseBlock: traverseBlock)
 
-            if let collectedValue = traverseBlock(subDict) {
+            if let collectedValue = traverseBlock(self, subDict) {
                 array.append(collectedValue)
             }
         }

